rustls/src/crypto/ring/tls12.rs

```diff
@@ -165,34 +165,20 @@ impl Tls12AeadAlgorithm for GcmAlgorithm {
 
     #[cfg(feature = "secret_extraction")]
     fn extract_keys(&self, key: &[u8], iv: &[u8], explicit: &[u8]) -> ConnectionTrafficSecrets {
-        let salt = {
```

Comment:
if people agree this is a reasonable thing to do, i can squash this into the earlier commits

Reply:
Seems reasonable. How about making it a function that does all three, taking three `const` arguments?

Reply:
Well, some need two and others need three. but then trying to take the three version and discard one of the arguments breaks type inference (and I think fixing that would mean writing out all the lengths, which I'd like to avoid).

This is what I ended up with:

```
#[cfg(feature = "secret_extraction")]
fn slices_to_arrays<const NK: usize, const NS: usize, const NI: usize>(
    k: &[u8],
    s: &[u8],
    i: &[u8],
) -> ([u8; NK], [u8; NS], [u8; NI]) {
    // these are guaranteed true because `ConnectionTrafficSecrets` items and
    // `key_len()` are in agreement.
    debug_assert_eq!(NK, k.len());
    debug_assert_eq!(NS, s.len());
    debug_assert_eq!(NI, i.len());
    (
        k.try_into().unwrap(),
        s.try_into().unwrap(),
        i.try_into().unwrap(),
    )
}
```

which works nicely in the 3-parameter case:

```
    #[cfg(feature = "secret_extraction")]
    fn extract_keys(&self, key: AeadKey, iv: Iv) -> ConnectionTrafficSecrets {
        let (key, salt, iv) = slices_to_arrays(key.as_ref(), &iv.0[..4], &iv.0[4..]);
        ConnectionTrafficSecrets::Aes256Gcm { key, salt, iv }
    }
```

but not for two:

```
    #[cfg(feature = "secret_extraction")]
    fn extract_keys(&self, key: AeadKey, iv: Iv) -> ConnectionTrafficSecrets {
        let (key, _, iv) = slices_to_arrays(key.as_ref(), &[], &iv.0);
        ConnectionTrafficSecrets::Chacha20Poly1305 { key, iv }
    }
```

Reply:
Squashed!